matrix-notifications.el

```diff
@@ -25,39 +25,112 @@
+                              (t (not enabled)))))
+    (unless key
+      ;; No room-specific setting yet: invert default
+      (setq key (or (elt aliases 0) id)
+            enabled (not matrix-client-enable-notifications)))
+    (map-put matrix-client-notifications-rooms key (list enabled fn))
+    ;; Save setting
+    ;; NOTE: Saving the setting now seems like the right thing to do, but it does cause a short
+    ;; delay while Emacs writes the custom file.  Is this the best way to handle this?
+    (customize-save-variable 'matrix-client-notifications-rooms matrix-client-notifications-rooms)
+    (message "Notifications for room %s: %s" key (cl-case enabled
+                                                   ('nil "off")
+                                                   (t "on")))))
+
 ;;;; Functions
 
-(defun matrix-client-notify (event-type data &rest rest)
-  "Run notify hooks and built-in notificataion for an event of EVENT-TYPE with DATA.
-Optional REST of args are also applied to hooks and function."
-  (when (and matrix-client-enable-notifications
-             (not matrix-client-initial-sync))
-    (run-hook-with-args 'matrix-client-notify-hook event-type data rest)
-    ;; Run built-in notification for this event type
-    (let ((fn (intern-soft (concat "matrix-client-notify-" event-type))))
-      (when (functionp fn)
-        (apply #'funcall fn data rest)))))
+(cl-defmethod matrix-client-notify ((room matrix-client-room) event-type data)
+  "When ROOM's notifications are enabled, run notify hooks and dispatcher for EVENT-TYPE with DATA."
+  ;; MAYBE: I think event-type is in DATA, so maybe we should get it there instead of passing it.
+  (unless matrix-client-initial-sync
+    (pcase-let* ((`(,key . (,enabled ,dispatch-fn)) (matrix-client--room-notification-settings room))
+                 (dispatch-fn (or dispatch-fn matrix-client-notify-dispatcher))
+                 (notify-this-event-p (cl-case key
+                                        ('nil  ;; No room-specific setting: use default
+                                         matrix-client-enable-notifications)
+                                        (t  ;; Room-specific setting
+                                         (or (when (functionp enabled)
+                                               (funcall enabled room event-type data))
+                                             enabled)))))
+      (when notify-this-event-p
+        (run-hook-with-args 'matrix-client-notify-hook room event-type data)
+        (funcall dispatch-fn room event-type data)))))
+
+(cl-defmethod matrix-client-notify-dispatch ((room matrix-client-room) event-type data)
+  "Dispatch DATA and REST to appropriate notification function for EVENT-TYPE.
+If no such function is defined, nothing happens."
+  (let ((fn (intern-soft (concat "matrix-client-notify-" event-type))))
+    (when (functionp fn)
```

Comment:
I would prefer an error here if fn wasn't a function, just so people would have an easier time debugging incorrect settings.

Reply:
Actually, this `matrix-client-notify-dispatch` function only dispatches to built-in functions, not user-defined ones.  Well, I mean, users could redefine the built-in ones, but that's not how this is designed to work.  For a custom notification function, users should change `matrix-client-notify-dispatcher` to their custom dispatch function.

Reply:
oh, I see, this sounds good to me then! I saw the function in customize and assumed that users might want to customize this themselves, but that makes a lot more sense :)

Reply:
Cool.  Users can also customize the dispatcher per-room.  And then there's the notify hook, so notifications can be easily run through multiple pipelines of notificationing...